backend/src/pong/pong.gateway.ts

```diff
@@ -10,117 +11,142 @@ import {
+		this.resetService()
 		console.log('\n!Socket is connected!\n');
-		this.pongService.gameIsFinished = false;
-		this.pongService.pointIsOver = false;
-		this.gameState = this.pongService.createNewGameState();
-		//  set score to win with requestGame info
-		this.gameState.scoreToWin = 10;
 	}
 
 	handleDisconnect(@ConnectedSocket() client: Socket) {
+		const cookie: string = jwtCookieFromHandshakeString(
+			client.handshake.headers.cookie,
+		);
+		const userId: number = this.authService.userIdFromCookieString(cookie);
+			// this.server.in( .roomName).emit('gameOver');
+			// gameSubscribers[data!.id] = undefined;
+			this.pongService.gameIsFinished = true;
 	}
 
-	@SubscribeMessage("joinGameRoom") 
+	createGameInstance(game: Game, client: Socket) {
+		const roomName = String(game.id);
+		gameSubscribers[game.id] = this.pongService.createNewGameState();
+		gameSubscribers[game.id].gameId = game.id;
+		gameSubscribers[game.id].scoreToWin = 10;
+		client.join(roomName);
+		gameSubscribers[game.id].roomName = roomName;
+	}
+
+	@SubscribeMessage('joinGameRoom')
 	joinGameRoom(@MessageBody() game: Game, @ConnectedSocket() client: Socket) {
+		console.log('CurretnGame: ', game);
 		const cookie: string = jwtCookieFromHandshakeString(
 			client.handshake.headers.cookie,
 		);
 		try {
 			const userId: number = this.authService.userIdFromCookieString(cookie);
-			if (userId === game.player_one.id) {
-				this.playerOneIsInGame = true;
-				this.playerOneId = game.player_one.id;
-			}
-			else if (userId === game.player_two.id) {
-				this.playerTwoIsInGame = true;
-			}
-			const roomName = this.buildUniqueRoomId(game);
-			client.join(roomName);
-			this.gameState.roomName = roomName;
-			if (this.playerOneIsInGame && this.playerTwoIsInGame) {
-				this.server.in(roomName).emit("GameCanStart");
+			if (game.id && game.player_one && game.player_two) {
+				if (!gameSubscribers[game.id]) {
+					this.createGameInstance(game, client);
+				} else {
+					client.join(gameSubscribers[game.id].roomName);
+				}
+				if (userId === game.player_one.id) {
+					gameSubscribers[game.id].playerOneIsInGame = true;
+					gameSubscribers[game.id].playerOneId = userId;
+				} else if (userId === game.player_two.id) {
+					gameSubscribers[game.id].playerTwoIsInGame = true;
+					gameSubscribers[game.id].playerTwoId = userId;
+				}
+				if (
+					!this.pongService.gameHasStarted &&
+					gameSubscribers[game.id].playerOneIsInGame &&
+					gameSubscribers[game.id].playerTwoIsInGame
+				) {
+					console.log('Emitting game start');
+					this.server
+						.in(gameSubscribers[game.id].roomName)
+						.emit('GameCanStart');
+					this.pongService.gameHasStarted = true;
+				}
 			}
 		} catch (e) {}
 	}
 
-	sendPositionOfElements(@MessageBody() gameState: GameState) {
-		this.server.in(this.gameState.roomName).emit('elementPositions', gameState);
+	sendPositionOfElements(gameState: GameState) {
+		this.server
+			.in(gameSubscribers[gameState.gameId].roomName)
```

Comment:
should this have a check if gameSubscribers[gameState.gameId] exists?
not sure if it's possible to get here after the room has been destroyed...
also not sure if `this.server.in("non_existant_room").emit()` throws or just does nothing, because if it just silently fails it's not a problem at all

Reply:
oh no.. never mind
than we still would have to protect accessing the `roomName` property of something undefined...

Reply:
In the "controller" on line 124, i check if the room exists, and only then proceed to call the methods that look "unprotected"